src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalescingAssignmentTests.cs

```diff
@@ -0,0 +1,1239 @@
+}
+");
+        }
+
+        [Fact]
+        public void UseBeforeAssignment()
+        {
+            CreateCompilation(@"
+public class C
+{
+    public static void Main()
+    {
+        C c;
+        c ??= new C();
+    }
+}").VerifyDiagnostics(new DiagnosticDescription[]
+{
+    // (7,9): error CS0165: Use of unassigned local variable 'c'
+    //         c ??= new C();
+    Diagnostic(ErrorCode.ERR_UseDefViolation, "c").WithArguments("c").WithLocation(7, 9)
+});
+        }
+
+        [Fact]
+        public void NonStaticInStaticContext()
+        {
+            CreateCompilation(@"
+public class C
+{
+    C P { get; set; }
+    public static void Main()
+    {
+        P ??= new C();
+    }
+}").VerifyDiagnostics(new DiagnosticDescription[]
+{
+    // (7,9): error CS0120: An object reference is required for the non-static field, method, or property 'C.P'
+    //         P ??= new C();
+    Diagnostic(ErrorCode.ERR_ObjectRequired, "P").WithArguments("C.P").WithLocation(7, 9)
+});
+        }
+
+        [Fact]
+        public void ValidRhsConversions()
+        {
+
+            CompileAndVerify(@"
+class C
+{
+    public static void Main()
+    {
+        // Implicit reference conversion
+        C c1 = null;
+        c1 ??= new D();
+
+        // Implicit user-defined conversion
+        C c2 = null;
+        c2 ??= new E();
+
+        // Implicit user-defined conversion to the underlying non-nullable type of lhs
+        int? i1 = null;
+        i1 ??= new C();
+    }
+
+    static public implicit operator C(E e) => null;
+    static public implicit operator int(C c) => 0;
+}
+class D : C {}
+class E {}
+").VerifyIL("C.Main()", @"
+{
+  // Code size       58 (0x3a)
+  .maxstack  1
+  .locals init (C V_0,
+                int? V_1)
+  IL_0000:  ldnull
+  IL_0001:  stloc.0
+  IL_0002:  ldloc.0
+  IL_0003:  brtrue.s   IL_000b
+  IL_0005:  newobj     ""D..ctor()""
+  IL_000a:  pop
+  IL_000b:  ldnull
+  IL_000c:  stloc.0
+  IL_000d:  ldloc.0
+  IL_000e:  brtrue.s   IL_001b
+  IL_0010:  newobj     ""E..ctor()""
+  IL_0015:  call       ""C C.op_Implicit(E)""
+  IL_001a:  pop
+  IL_001b:  ldloca.s   V_1
+  IL_001d:  initobj    ""int?""
+  IL_0023:  ldloc.1
+  IL_0024:  stloc.1
+  IL_0025:  ldloca.s   V_1
+  IL_0027:  call       ""bool int?.HasValue.get""
+  IL_002c:  brtrue.s   IL_0039
+  IL_002e:  newobj     ""C..ctor()""
+  IL_0033:  call       ""int C.op_Implicit(C)""
+  IL_0038:  pop
+  IL_0039:  ret
+}
+");
+        }
+
+        [Fact]
+        public void InvalidRhsConversions()
+        {
+            CreateCompilation(@"
+class C
+{
+    public void M()
+    {
+        // Explicit numeric conversion
+        int? i1 = null;
+        i1 ??= 1.0;
+
+        // Explicit reference conversion
+        D d1 = null;
+        d1 ??= new C();
+
+        // Explicit user-defined conversion
+        C c1 = null;
+        c1 ??= new E();
+
+        // No conversion between lhs and rhs
+        C c2 = null;
+        c2 ??= new F();
+    }
```

Comment:
>} [](start = 4, length = 1)

Consider testing `derived ??= base`.

Reply:
That's the second error here

---
In reply to: [199837406](https://github.com/dotnet/roslyn/pull/28253#discussion_r199837406) [](ancestors = 199837406)